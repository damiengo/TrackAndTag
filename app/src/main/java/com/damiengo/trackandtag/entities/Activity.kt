package com.damiengo.trackandtag.entities

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import java.time.LocalDateTime

@Entity
data class Activity(
    @PrimaryKey val activityId: Long,
    @ColumnInfo(name = "title") val title: String?,
    @ColumnInfo(name = "description") val description: String?,
    @ColumnInfo(name = "number") val number: Float?,
    @ColumnInfo(name = "made_at") val madeAt: LocalDateTime?,
    @ColumnInfo(name = "created_at") val createdAt: LocalDateTime?,
    @ColumnInfo(name = "updated_at") val updatedAt: LocalDateTime?
)
