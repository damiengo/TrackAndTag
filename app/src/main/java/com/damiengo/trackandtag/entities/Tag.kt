package com.damiengo.trackandtag.entities

import android.os.Build
import androidx.annotation.RequiresApi
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import java.time.LocalDateTime

@Entity
data class Tag(
    @PrimaryKey(autoGenerate = true) var tagId: Long,
    @ColumnInfo(name = "text") val title: String?,
    @ColumnInfo(name = "created_at") val createdAt: LocalDateTime?,
    @ColumnInfo(name = "updated_at") val updatedAt: LocalDateTime?
) {
    @RequiresApi(Build.VERSION_CODES.O)
    @Ignore
    constructor(text: String = "",
                createdAt: LocalDateTime? = LocalDateTime.now()) : this(0, text, createdAt, null)
}
