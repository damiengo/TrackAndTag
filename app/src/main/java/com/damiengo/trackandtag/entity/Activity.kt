package com.damiengo.trackandtag.entity

import android.os.Build
import androidx.annotation.RequiresApi
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey

@Entity
data class Activity(
    @PrimaryKey(autoGenerate = true) var activityId: Long,
    @ColumnInfo val title: String?,
    @ColumnInfo val description: String?,
    @ColumnInfo val number: Float?,
    @ColumnInfo val madeAt: Long,
    @ColumnInfo val createdAt: Long,
    @ColumnInfo val updatedAt: Long
) {
    @RequiresApi(Build.VERSION_CODES.O)
    @Ignore
    constructor(title: String = "",
                description: String = "",
                number: Float? = 0f,
                madeAt: Long = System.currentTimeMillis(),
                createdAt: Long = System.currentTimeMillis()) : this(0, title, description, number, madeAt, createdAt, 0)
}