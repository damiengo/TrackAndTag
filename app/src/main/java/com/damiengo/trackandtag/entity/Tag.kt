package com.damiengo.trackandtag.entity

import android.os.Build
import androidx.annotation.RequiresApi
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey

@Entity
data class Tag(
    @PrimaryKey(autoGenerate = true) var tagId: Long,
    @ColumnInfo val text: String?,
    @ColumnInfo val createdAt: Long,
    @ColumnInfo val updatedAt: Long
) {
    @RequiresApi(Build.VERSION_CODES.O)
    @Ignore
    constructor(text: String = "",
                createdAt: Long = System.currentTimeMillis()) : this(0, text, createdAt, 0)
}
