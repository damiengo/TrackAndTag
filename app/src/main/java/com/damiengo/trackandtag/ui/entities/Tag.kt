package com.damiengo.trackandtag.ui.entities

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import java.time.LocalDateTime

@Entity
data class Tag(
    @PrimaryKey val uid: Int,
    @ColumnInfo(name = "text") val title: String?,
    @ColumnInfo(name = "created_at") val createdAt: LocalDateTime?,
    @ColumnInfo(name = "updated_at") val updatedAt: LocalDateTime?
)
