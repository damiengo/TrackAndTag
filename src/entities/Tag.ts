import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  OneToMany, 
  Unique
} from 'typeorm';

import { ActivityTag } from './ActivityTag'

@Entity('tag')
@Unique(['label'])
export class Tag {

  @PrimaryGeneratedColumn() id: number

  @Column('text')
  label: string

  @Column('text', { nullable: true })
  description: string | null

  @Column({ nullable: true })
  unit: number | undefined

  @Column()
  createdAt: number

  @Column({ nullable: true })
  updatedAt: number

  @OneToMany(/* istanbul ignore next */() => ActivityTag, /* istanbul ignore next */activityTag => activityTag.tag)
  activityTags: ActivityTag[]

  activities = () => this.activityTags.map((at: ActivityTag) => at.activity)

}