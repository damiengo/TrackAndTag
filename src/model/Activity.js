import { Model } from '@nozbe/watermelondb'
import { field, date, text, lazy } from '@nozbe/watermelondb/decorators'

export default class Activity extends Model {

  static table = 'activities'

  @text('title') title
  @text('description') description
  @field('number') number
  @date('made_at') madeAt
  @date('created_at') createdAt
  @date('updated_at') updatedAt

  static associations = {
    activity_tags: { type: 'has_many', foreignKey: 'activity_id' },
  }

  @lazy
  tags = this.collections
    .get('tags')
    .query(Q.on('activity_tags', 'activity_id', this.id));

}