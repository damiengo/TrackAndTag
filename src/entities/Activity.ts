import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  ManyToMany, 
  JoinTable,
  OneToMany, 
  Index
} from 'typeorm';

import { ActivityTag } from './ActivityTag'
import { Tag } from './Tag'

@Entity('activity')
export class Activity {

  @PrimaryGeneratedColumn() 
  id: number

  @Column('text') 
  title: string

  @Column('text', { nullable: true })
  description: string | undefined

  @Column('real')
  @Index()
  quantity: number

  @Column()
  @Index()
  madeAt: number

  @Column()
  @Index()
  createdAt: number

  @Column({ nullable: true })
  @Index()
  updatedAt: number | undefined

  @OneToMany(/* istanbul ignore next */() => ActivityTag, 
             /* istanbul ignore next */activityTag => activityTag.activity)
  activityTags: ActivityTag[]

  tagsLabels: string

  getTags = () : Tag[] => this.activityTags.map((at: ActivityTag) => at.tag)
  getTagsLabels = () : string[] => this.getTags().map((tag: Tag) => tag.label)

  addTags = (tags: Tag[]) => tags.map((tag: Tag) => {
    const at = new ActivityTag()
    at.activity = this
    at.tag = tag
    if(this.activityTags == null) {
      this.activityTags = []
    }
    this.activityTags.push(at)
  })

}