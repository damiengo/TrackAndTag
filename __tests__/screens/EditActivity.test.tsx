import 'react-native';
import React from 'react';
import EditActivityScreen from '../../src/screens/EditActivity';

import { Button, TextInput, Alert } from 'react-native'

import { create, act } from 'react-test-renderer';
jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');

import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

jest.mock('@react-navigation/native');
jest.mock('react-native-sqlite-storage');

import { Activity } from '../../src/entities/Activity';
import { Tag } from '../../src/entities/Tag';

describe('activity', () => {

  it('add', async () => {
      const route = {}
      const navigation = {
          setOptions: jest.fn(),
          goBack: jest.fn()
      }
      const wrapper = shallow(<EditActivityScreen route={route} navigation={navigation} />)
      wrapper.find({'testID': 'input_tags'}).simulate('changeText', 'tag7 tag8')
      wrapper.find({'testID': 'input_title'}).simulate('changeText', 'Title added 1')
      wrapper.find({'testID': 'input_description'}).simulate('changeText', 'Description added 2')
      wrapper.find({'testID': 'input_number'}).simulate('changeText', 88)
      wrapper.find({'testID': 'input_date'}).simulate('focus')
      wrapper.find({'testID': 'dateTimePicker'}).simulate('change')
      wrapper.find({'testID': 'button_save'}).simulate('press')
      await Promise.resolve();
      expect(navigation.goBack.mock.calls.length).toEqual(1)
      //const activities = await getActivities()
      const activities: Activity[] = []
      if(activities && activities.length == 1) {
        const savedActivity = activities[0]
        expect(savedActivity.title).toBe('Title added 1')
        expect(savedActivity.description).toBe('Description added 2')
        expect(savedActivity.quantity).toBe(88)
        const tags = savedActivity.getTags()
        if(tags && tags.length == 2) {
          const tag1 = tags[0]
          const tag2 = tags[1]
          expect(tag1.label).toBe('tag7')
          expect(tag2.label).toBe('tag8')
        }
        else {
          fail('No tags saved')
        }
      }
      else {
        fail('No activity saved')
      }
  });

  it('edit', async () => {
      jest.clearAllMocks()
      const tag1 = new Tag()
      tag1.label = 'tag1'
      const tag2 = new Tag()
      tag2.label = 'tag2'
      const activity = new Activity()
      activity.id = 1
      activity.title = 'Title item to edit'
      activity.description = 'Description to edit'
      activity.quantity = 5
      activity.madeAt = (new Date()).getTime()
      activity.createdAt = (new Date()).getTime()
      activity.addTags([ tag1, tag2 ])
      const route = {
        params: {
          item: activity
        }
      }
      const navigation = {
          setOptions: jest.fn(),
          goBack: jest.fn()
      }
      const wrapper = shallow(<EditActivityScreen route={route} navigation={navigation} />)
      wrapper.find({'testID': 'input_description'}).simulate('changeText', 'Description edited 8')
      wrapper.find({'testID': 'input_number'}).simulate('changeText', 9)
      wrapper.find({'testID': 'button_save'}).simulate('press')
      await Promise.resolve();
      expect(navigation.goBack.mock.calls.length).toEqual(1)
      expect(wrapper).toMatchSnapshot()
      //const activities = await getActivities()
      const activities: Activity[] = []
      if(activities && activities.length == 1) {
        const savedActivity = activities[0]
        expect(savedActivity.title).toBe('Title item to edit')
        expect(savedActivity.description).toBe('Description edited 8')
        expect(savedActivity.quantity).toBe(9)
        const tags = savedActivity.getTags()
        if(tags && tags.length == 2) {
          const tag1 = tags[0]
          const tag2 = tags[1]
          expect(tag1.label).toBe('tag1')
          expect(tag2.label).toBe('tag2')
        }
        else {
          fail('No tags saved')
        }
      }
  });

  it('delete', async () => {
      const activity = new Activity()
      activity.id = 1
      activity.title = 'Title item to delete'
      activity.description = 'Description to delete'
      activity.quantity = 5
      activity.madeAt = (new Date()).getTime()
      activity.createdAt = (new Date()).getTime()
      activity.activityTags = []
      const route = {
        params: {
          item: activity
        }
      }
      const navigation = {
          setOptions: jest.fn(),
          goBack: jest.fn()
      }
      const wrapper = shallow(<EditActivityScreen route={route} navigation={navigation} />)
      wrapper.find({'testID': 'button_delete'}).simulate('press')
      await Promise.resolve()
      expect(wrapper).toMatchSnapshot()
  });
});