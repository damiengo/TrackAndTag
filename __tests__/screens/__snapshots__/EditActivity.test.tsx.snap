// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activity delete 1`] = `ShallowWrapper {}`;

exports[`activity edit 1`] = `ShallowWrapper {}`;
